File: agent/library/src/main/java/org/matilda/services/plugins/JavaLoader.java
package org.matilda.services.plugins;

import javax.inject.Inject;
import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.util.HashMap;
import java.util.Map;
import java.util.jar.JarEntry;
import java.util.jar.JarInputStream;

public class JavaLoader {
    @Inject
    JavaLoader() {}


    public ClassLoader load(byte[] jarBytes) throws IOException {
        Map<String, byte[]> classes = new HashMap<>();

        try (JarInputStream jarInputStream = new JarInputStream(new ByteArrayInputStream(jarBytes))) {
            JarEntry jarEntry;
            while ((jarEntry = jarInputStream.getNextJarEntry()) != null) {
                String entryName = jarEntry.getName();
                if (entryName.endsWith(".class")) {
                    // Extract the class name from the entry name
                    String className = entryName.replace('/', '.').replaceAll("\\.class$", "");
                    byte[] classData = readFully(jarInputStream);
                    classes.put(className, classData);
                }
            }
        }

        return new InMemoryClassLoader(classes);
    }

    private byte[] readFully(InputStream inputStream) throws IOException {
        byte[] buffer = new byte[4096];
        ByteArrayOutputStream byteArrayOutputStream = new ByteArrayOutputStream();
        int bytesRead;
        while ((bytesRead = inputStream.read(buffer)) != -1) {
            byteArrayOutputStream.write(buffer, 0, bytesRead);
        }
        return byteArrayOutputStream.toByteArray();
    }
}

File: agent/library/src/main/java/org/matilda/services/plugins/JavaLoader.kt
package org.matilda.services.plugins

import java.io.ByteArrayInputStream
import java.io.ByteArrayOutputStream
import java.io.IOException
import java.io.InputStream
import java.util.jar.JarEntry
import java.util.jar.JarInputStream
import javax.inject.Inject

class JavaLoader @Inject internal constructor() {
    fun load(jarBytes: ByteArray): ClassLoader {
        val classes: MutableMap<String, ByteArray> = HashMap()
        JarInputStream(ByteArrayInputStream(jarBytes)).use { jarInputStream ->
            var jarEntry: JarEntry
            while (jarInputStream.nextJarEntry.also { jarEntry = it } != null) {
                val entryName = jarEntry.name
                if (entryName.endsWith(".class")) {
                    // Extract the class name from the entry name
                    val className = entryName.replace('/', '.').replace("\\.class$".toRegex(), "")
                    val classData = readFully(jarInputStream)
                    classes[className] = classData
                }
            }
        }
        return InMemoryClassLoader(classes)
    }

    @Throws(IOException::class)
    private fun readFully(inputStream: InputStream): ByteArray {
        val buffer = ByteArray(4096)
        val byteArrayOutputStream = ByteArrayOutputStream()
        var bytesRead: Int
        while (inputStream.read(buffer).also { bytesRead = it } != -1) {
            byteArrayOutputStream.write(buffer, 0, bytesRead)
        }
        return byteArrayOutputStream.toByteArray()
    }
}
